File: src/main/java/io/github/ajoz/workshop/fp/java/tools/seq/ArrayFlow.java
package io.github.ajoz.workshop.fp.java.tools.seq;

import io.github.ajoz.workshop.fp.java.tools.Try;

import java.util.NoSuchElementException;

public final class ArrayFlow<A> implements Flow<A> {
    private final A[] array;
    private int current;

    @SafeVarargs
    ArrayFlow(final A... array) {
        this.array = array;
    }

    @Override
    public Try<A> next() {
        if (array.length == 0)
            return Try.failure(new NoSuchElementException("No elements in this Flow!"));

        if (current >= array.length)
            return Try.failure(new NoSuchElementException("No more elements in this Flow!"));

        final Try<A> next = Try.success(array[current]);
        current++;
        return next;
    }
}

File: src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/tools/flow/ArrayFlow.kt
@file:Suppress("unused")

package io.github.ajoz.workshop.fp.kotlin.tools.flow

import io.github.ajoz.workshop.fp.kotlin.tools.control.Try
import java.util.NoSuchElementException

/**
 * Creates a [Flow] from a given array of values.
 *
 * Each element of the given array will be used by the Flow. In case of the
 * array being empty a [Try.Failure] will be send to the downstream. In case of
 * sending all the available elements downstream this Flow will only return
 * [Try.Failure].
 *
 * @param A the type of the returned flow.
 * @param[values] array of values that should be lazily returned in the form of
 * a flow.
 * @return instance of a Flow.
 */
fun <A> Flow.Companion.of(vararg values: A): Flow<A> {
    class ArrayFlow(val array: Array<out A>) : Flow<A> {
        private var current: Int = 0

        override fun next(): Try<A> {
            if (array.isEmpty())
                return Try.Failure(NoSuchElementException("No elements in this Flow!"))

            if (current >= array.size)
                return Try.Failure(NoSuchElementException("No more elements in this Flow!"))

            val next: Try<A> = Try.Success(array[current])
            current++
            return next
        }
    }

    return ArrayFlow(values)
}


